src/simulators/statevector/qvintrin_avx.hpp

```diff
@@ -0,0 +1,955 @@
+    const areg_t<1ULL << N> &inds,
+    const areg_t<N>& qregs
+){
+
+  const int PERM_D_Q0 = 2 * 64 + 3 * 16 + 0 * 4 + 1 * 1;
+  const int PERM_D_Q1 = 1 * 64 + 0 * 16 + 3 * 4 + 2 * 1;
+
+  __m256d real_ret, imag_ret, real_ret1, imag_ret1;
+  __m256d vreals[1ULL << N], vimags[1ULL << N];
+  __m256d tmp0, tmp1;
+
+  const double * dmat = reinterpret_cast<const double *>(mat.data());
+
+  for(unsigned i = 0; i < (1ULL << N); i += 2){
+    auto idx = inds[i];
+    _mm_load_twoarray_complex(reals[idx], imags[idx], vreals[i], vimags[i]);
+    if (qregs[0] == 0) {
+      vreals[i + 1] = _mm256_permute4x64_pd(vreals[i], PERM_D_Q0);
+      vimags[i + 1] = _mm256_permute4x64_pd(vimags[i], PERM_D_Q0);
+    } else {
+      vreals[i + 1] = _mm256_permute4x64_pd(vreals[i], PERM_D_Q1);
+      vimags[i + 1] = _mm256_permute4x64_pd(vimags[i], PERM_D_Q1);
+    }
+  }
+
+  size_t midx = 0;
+  for(size_t i = 0; i < (1ULL << N); i += 2){
+    auto idx = inds[i];
+    _mm_complex_inner_product<double>((1ULL << N), vreals, vimags, (&dmat[midx]), real_ret, imag_ret, tmp0, tmp1);
+    midx += (1ULL << (N + 1));
+
+    _mm_complex_inner_product<double>((1ULL << N), vreals, vimags, (&dmat[midx]), real_ret1, imag_ret1, tmp0, tmp1);
+    midx += (1ULL << (N + 1));
+
+    if (qregs[0] == 0) {
+      real_ret1 = _mm256_permute4x64_pd(real_ret1, PERM_D_Q0);
+      imag_ret1 = _mm256_permute4x64_pd(imag_ret1, PERM_D_Q0);
+      real_ret = _mm256_blend_pd(real_ret, real_ret1, 0b1010);
+      imag_ret = _mm256_blend_pd(imag_ret, imag_ret1, 0b1010);
+    } else {
+      real_ret1 = _mm256_permute4x64_pd(real_ret1, PERM_D_Q1);
+      imag_ret1 = _mm256_permute4x64_pd(imag_ret1, PERM_D_Q1);
+      real_ret = _mm256_blend_pd(real_ret, real_ret1, 0b1100);
+      imag_ret = _mm256_blend_pd(imag_ret, imag_ret1, 0b1100);
+    }
+    _mm_store_twoarray_complex(real_ret, imag_ret, reals[idx], imags[idx]);
+  }
+}
+
+template<size_t N>
+inline void _apply_matrix_double_avx(
+    RealVectorView<double>& reals,
+    ImaginaryVectorView<double>& imags,
+    const QV::cvector_t<double>& mat,
+    const areg_t<1ULL << N> &inds,
+    const areg_t<N> qregs
+){
+
+  __m256d real_ret, imag_ret;
+  __m256d vreals[1ULL << N], vimags[1ULL << N];
+  __m256d tmp0, tmp1;
+
+  const double * dmat = reinterpret_cast<const double *>(mat.data());
+
+  for(unsigned i = 0; i < (1ULL << N); ++i){
+    auto idx = inds[i];
+    _mm_load_twoarray_complex(reals[idx], imags[idx], vreals[i], vimags[i]);
+  }
+
+  unsigned midx = 0;
+  for(unsigned i = 0; i < (1ULL << N); ++i){
+    auto idx = inds[i];
+    _mm_complex_inner_product<double>((1ULL << N), vreals, vimags, (&dmat[midx]), real_ret, imag_ret, tmp0, tmp1);
+    midx += (1ULL << (N + 1));
+    _mm_store_twoarray_complex(real_ret, imag_ret, reals[idx], imags[idx]);
+  }
+}
+
+template<typename list_t>
+uint64_t _index0(const list_t &qubits_sorted, const uint64_t k){
+  uint64_t lowbits, retval = k;
+  static constexpr std::array<uint64_t, 64> _MASKS = {
```

Comment:
This should be included from separate file so we aren't duplicating the index functions in multiple files (See `indexes.hpp` in #718)